wrapper/src/main/java/software/amazon/jdbc/plugin/customendpoint/CustomEndpointMonitorImpl.java

```diff
@@ -0,0 +1,238 @@
+          List<String> endpointURLs = endpoints.stream().map(DBClusterEndpoint::endpoint).collect(Collectors.toList());
+          LOGGER.warning(
+              Messages.get("CustomEndpointMonitorImpl.unexpectedNumberOfEndpoints",
+                  new Object[] {
+                      this.endpointIdentifier,
+                      this.region.id(),
+                      endpoints.size(),
+                      endpointURLs
+                  }
+              ));
+
+          TimeUnit.NANOSECONDS.sleep(this.refreshRateNano);
+          continue;
+        }
+
+        CustomEndpointInfo endpointInfo = CustomEndpointInfo.fromDBClusterEndpoint(endpoints.get(0));
+        CustomEndpointInfo cachedEndpointInfo = customEndpointInfoCache.get(this.customEndpointHostSpec.getHost());
+        if (cachedEndpointInfo != null && cachedEndpointInfo.equals(endpointInfo)) {
+          long elapsedTime = System.nanoTime() - start;
+          long sleepDuration = Math.min(0, this.refreshRateNano - elapsedTime);
+          TimeUnit.NANOSECONDS.sleep(sleepDuration);
+          continue;
+        }
+
+        LOGGER.fine(
+            Messages.get(
+                "CustomEndpointMonitorImpl.detectedChangeInCustomEndpointInfo",
+                new Object[]{ this.customEndpointHostSpec.getHost(), endpointInfo }));
+
+        // The custom endpoint info has changed, so we need to update the plugin service info cache.
+        customEndpointInfoCache.put(
+            this.customEndpointHostSpec.getHost(), endpointInfo, CUSTOM_ENDPOINT_INFO_EXPIRATION_NANO);
+        this.pluginService.setInfo(this.customEndpointHostSpec.getHost(), endpointInfo, true);
+        this.infoChangedCounter.inc();
+
+        long elapsedTime = System.nanoTime() - start;
+        long sleepDuration = Math.min(0, this.refreshRateNano - elapsedTime);
+        TimeUnit.NANOSECONDS.sleep(sleepDuration);
+      }
+    } catch (InterruptedException e) {
+      LOGGER.info(Messages.get("CustomEndpointMonitorImpl.interrupted", new Object[]{ this.customEndpointHostSpec }));
+      Thread.currentThread().interrupt();
+    } catch (Exception e) {
+      LOGGER.log(Level.SEVERE,
+          Messages.get("CustomEndpointMonitorImpl.exception", new Object[]{this.customEndpointHostSpec.getHost()}), e);
+    } finally {
+      this.rdsClient.close();
+      LOGGER.fine(
+          Messages.get("CustomEndpointMonitorImpl.stoppedMonitor", new Object[]{ this.customEndpointHostSpec }));
+    }
+  }
+
+  @Override
+  public boolean shouldDispose() {
+    return true;
+  }
+
+  /**
+   * Stops the custom endpoint monitor.
+   */
+  @Override
+  public void close() {
+    LOGGER.fine(
+        Messages.get(
+            "CustomEndpointMonitorImpl.stoppingMonitor",
+            new Object[]{ this.customEndpointHostSpec.getHost() }));
+
+    this.stop.set(true);
+
+    try {
+      // TODO: the run loop takes 30s by default but could take more depending on the user setting. Should we keep 30s
```

Comment:
I think we can wait less than 30s. Most of the time the thread sleeps that can be interrupted by `this.monitorExecutor.shutdownNow()`. Any interaction with RDS API takes seconds to complete. I'd wait just 5s.